cmd/flat-network-controller/app/controller_test.go

```diff
@@ -0,0 +1,167 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	v1 "k8s.io/arktos-ext/pkg/apis/arktosextensions/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	coremock "k8s.io/client-go/testing"
+	api "k8s.io/kubernetes/pkg/apis/core"
+
+	//"k8s.io/client-go/kubernetes/fake"
+	fakearktosv1 "k8s.io/arktos-ext/pkg/generated/clientset/versioned/fake"
+
+	"testing"
+)
+
+func TestManageFlatNetwork(t *testing.T) {
+	tcs := []struct {
+		desc           string
+		input          *v1.Network
+		svcResp        *corev1.Service
+		svcRespError   error
+		netRespError   error
+		expectingError bool
+	}{
+		{
+			desc: "happy path",
+			input: &v1.Network{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "test-ne",
+					Tenant: "test-te",
+				},
+				Spec: v1.NetworkSpec{
+					Type: "flat",
+				},
+				Status: v1.NetworkStatus{},
+			},
+			svcResp: &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{},
+				Spec: corev1.ServiceSpec{
+					ClusterIP: "11.22.33.44",
+					Type:      "ClusterIP",
+				},
+				Status: corev1.ServiceStatus{},
+			},
+			expectingError: false,
+		},
+		{
+			desc: "dns service unable to create",
+			input: &v1.Network{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "test-ne",
+					Tenant: "test-te",
+				},
+				Spec: v1.NetworkSpec{
+					Type: "flat",
+				},
+				Status: v1.NetworkStatus{},
+			},
+			svcResp:        &corev1.Service{},
+			svcRespError:   fmt.Errorf("fake test error: failed to create DNS service"),
+			expectingError: true,
+		},
+		{
+			desc: "dns service already exists",
+			input: &v1.Network{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "test-ne",
+					Tenant: "test-te",
+				},
+				Spec: v1.NetworkSpec{
+					Type: "flat",
+				},
+				Status: v1.NetworkStatus{},
+			},
+			svcResp: &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Tenant:    "test-te",
+					Namespace: "kube-system",
+					Name:      "kube-dns-test-ne",
+				},
+				Spec: corev1.ServiceSpec{
+					ClusterIP: "11.22.33.44",
+					Type:      "ClusterIP",
+				},
+				Status: corev1.ServiceStatus{},
+			},
+			svcRespError:   errors.NewAlreadyExists(api.Resource("service"), "kube-dns-test-ne"),
+			expectingError: false,
+		},
+		{
+			desc: "network status update failed",
+			input: &v1.Network{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "test-ne",
+					Tenant: "test-te",
+				},
+				Spec: v1.NetworkSpec{
+					Type: "flat",
+				},
+				Status: v1.NetworkStatus{},
+			},
+			svcResp: &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{},
+				Spec: corev1.ServiceSpec{
+					ClusterIP: "11.22.33.44",
+					Type:      "ClusterIP",
+				},
+				Status: corev1.ServiceStatus{},
+			},
+			netRespError:   fmt.Errorf("fake test error: failed to  update network status"),
+			expectingError: true,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.desc, func(t *testing.T) {
+			netClient := fakearktosv1.NewSimpleClientset(tc.input)
```

Comment:
As we have a field called "desc" in test case, let's include it in the error log, for the convenience of debugging. 

For ex, change line 143 to :
t.Fatalf("%v: unexpected update", t.desc)

Reply:
subtest would automatically put desc (e.g. "dns service already exists") in test case summary like
```=== RUN   TestManageFlatNetwork/dns_service_already_exists```